framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,588 @@
+import org.checkerframework.framework.util.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
+
+import annotations.Annotation;
+import annotations.el.AClass;
+import annotations.el.AField;
+import annotations.el.AMethod;
+import annotations.el.AScene;
+import annotations.el.AnnotationDef;
+import annotations.el.DefException;
+import annotations.field.AnnotationFieldType;
+import annotations.field.ArrayAFT;
+import annotations.field.BasicAFT;
+import annotations.field.ScalarAFT;
+import annotations.io.IndexFileParser;
+import annotations.io.IndexFileWriter;
+import annotations.util.JVMNames;
+
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Attribute.Array;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.TypeSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Type.ArrayType;
+import com.sun.tools.javac.code.Type.ClassType;
+
+/**
+ * This class writes inferred types for fields, method return types and method
+ * parameters to a .jaif file.  Calling an update* method
+ * ({@link SignatureInferenceScenes#updateFieldTypeInScene updateFieldTypeInScene},
+ * {@link SignatureInferenceScenes#updateMethodParameterTypeInScene updateMethodParameterTypeInScene} or
+ * {@link SignatureInferenceScenes#updateMethodReturnTypeInScene updateMethodReturnTypeInScene}) 
+ * reads the currently-stored type, if any, and replaces it by the LUB of
+ * it and the update method's argument.
+ *
+ * This class does not perform inference for elements that have explicit
+ * annotations - an explicitly annotated field type, method return type or
+ * method parameter type will not have its respective inferred type written
+ * into a .jaif file.
+ *
+ * @author pbsf
+ */
+public class SignatureInferenceScenes {
+
+    /**
+     * Path to where .jaif files will be written to and read from.
+     * This path is relative to where the CF's javac command is executed.
+     */
+    public final static String JAIF_FILES_PATH = "build/jaif-files/";
+
+    /** Maps file paths (Strings) to Scenes. */
+    private static Map<String, AScene> scenes = new HashMap<String, AScene>();
+
+    /** Set containing all Scenes that were modified in the current ClassTree.
+     * Modifying a Scene means adding (or changing) a type annotation for a
+     * field, method return type or method parameter type in the Scene.
+     * (Scenes are modified by the method
+     * {@link SignatureInferenceScenes#updateAnnotationSetInScene updateAnnotationSetInScene)}
+     */
+    private static Set<String> modifiedScenes = new HashSet<String>();
+
+    /**
+     * Returns the Scene related to a .jaif file path passed as input.
```

Comment:
No, it is not possible. If a Scene is updated to have a different type in a certain location, and then its written into a .jaif file, the Scene will still contain that modification, and it will be retrieved by that operation. A .jaif file cannot be modified without modifying the Scene first. So the Scene will always be updated, and once written into a .jaif file, the file will also be updated.